server/main_server/src/main.cpp

```diff
@@ -0,0 +1,1342 @@
+        // to fork(). To avoid potential issues, we build up the argument list before
+        // doing the fork-exec.
+
+        auto binary = (irods::get_irods_sbin_directory() / "irodsAgent").string();
+        std::string hn_shm_name{irods::experimental::net::hostname_cache::shared_memory_name()};
+        std::string dns_shm_name{irods::experimental::net::dns_cache::shared_memory_name()};
+
+        std::vector<char*> args{binary.data(), hn_shm_name.data(), dns_shm_name.data()};
+
+        if (_write_to_stdout) {
+            args.push_back("--stdout");
+        }
+
+        if (_enable_test_mode) {
+            args.push_back("-t");
+        }
+
+        args.push_back(nullptr);
+
+        g_pid_af = fork();
+
+        if (0 == g_pid_af) {
+            execv(args[0], args.data());
+
+            // If execv() fails, the POSIX standard recommends using _exit() instead of exit() to avoid
+            // flushing stdio buffers and handlers registered by the parent.
+            //
+            // In the case of C++, this is necessary to avoid triggering destructors. Triggering a destructor
+            // could result in assertions made by the struct/class being violatied. For some data types,
+            // violating an assertion results in program termination (i.e. SIGABRT).
+            _exit(1);
+        }
+        else if (-1 == g_pid_af) {
+            g_pid_af = 0;
+            log_server::error("{}: Could not launch agent factory.", __func__);
+            return false;
+        }
+
+        log_server::info("{}: Agent Factory PID = [{}].", __func__, g_pid_af);
+        return true;
+    } // launch_agent_factory
+
+    auto handle_configuration_reload(bool _write_to_stdout, bool _enable_test_mode) -> void
+    {
+        irods::at_scope_exit reset_reload_flag{[] { g_reload_config = 0; }};
+
+        log_server::info("{}: Received configuration reload instruction. Reloading configuration.", __func__);
+
+        if (!validate_configuration()) {
+            log_server::error("{}: Invalid configuration. Continuing to run with previous configuration.", __func__);
+            return;
+        }
+
+        if (g_pid_ds > 0) {
+            log_server::info("{}: Sending SIGTERM to delay server.", __func__);
+            kill(g_pid_ds, SIGTERM);
+        }
+
+        log_server::info("{}: Sending SIGQUIT to agent factory.", __func__);
+        kill(g_pid_af, SIGQUIT);
+
+        // Reset this variable so that the delay server migration logic can handle
+        // the relaunching of the delay server for us.
+        g_pid_ds = 0;
+
+        try {
+            log_server::info("{}: Reloading configuration for main server process.", __func__);
+            // TODO Loading the config files must be treated as an atomic operation. If either operation
+            // fails, the server may be in an invalid state.
+            irods::server_properties::instance().reload();
+            irods::environment_properties::instance().capture();
```

Comment:
This MUST be treated as an atomic operation. Reloading the config MUST stay in sync. If an error occurs, both configs (in memory) must use the old/existing configs.

Reply:
See if nginx and/or apache httpd handle this. I'm mostly interested in the failure cases and how they handle them.

Reply:
I'm happy with the latest implementation for this.

The implementation has been reviewed with the team.

Resolving this.